### ISSUE TITLE
`keyof typeof` causes a false-positive recursive type error

### ISSUE THREAD

**Issue by Starwort** [microsoft/TypeScript#55631]
### 🔎 Search Terms

"keyof typeof", "keyof typeof satisfies", "auto", "satisfies recursive"

### 🕗 Version & Regression Information

- This is the behavior in every version I tried (4.9.5, nightly 20230904), and I reviewed the FAQ for entries about satisfied and recursive type definitions

### ⏯ Playground Link

https://www.typescriptlang.org/play?ts=5.3.0-dev.20230904#code/JYOwLgpgTgZghgYwgAgEoRtAzgFQPYDSEAnsgN4BQy1yADlHrQFzIDWJeMyYxtEnyACIBBHMIDcFAL4UKAejlsOXHnwEixyYFmQATbQCs8oMMhgMAtmgzZ8RYuO4ALFO2ICsTvAFcANruRoOCwUMBdkBGAoBD84KD0IPhBdCBAEYgoEPBAsUw1hZABecioaGDw8FjJ6RhYAcgAjOLqpABpS6iaoKprmZDryvBb2qWQsODBtGGAIHXQsqF0AHjgQYlbrTChcQhIAPkkKE2h4JCE8WfAcKGBaEpo6Bj63AVV+LnyAJkkZTOzcoSiYSfFjzPCLFZrDaCC5YK43Wh7Ir3GgKPSwkB1UxgBHOFC+EwbVYBAC0yAA7uCQsgyW8EC4EKwdFhvAgkFgsDA-L5SCksLRgJBkA0IKAAOYUhggMUdMwVHpPepdYbSIA

### 💻 Code

```ts
interface RefersToKey {
    prop: keyof typeof DATA;
}

// keyof typeof DATA is disjoint from RefersToKey; the keyof should erase the circular dependency
const DATA = {
    foo: {prop: 'bar'},
    bar: {prop: 'foo'},
} satisfies Record<any, RefersToKey>;

interface DoesntTrip {
    prop: keyof typeof DATA2;
}

const DATA2: Record<any, DoesntTrip> = {
    // doesn't trip the lint, and - worse - typechecks successfully despite being wrong
    foo: {prop: 'bar'},
}
```

### 🙁 Actual behavior

TS2502 is issued on `prop` of `RefersToKey`, complaining that it is a circular dependency. No lint is issued on `DoesntTrip`, indicating that the compiler is capable of understanding that this is in fact not a circular dependency

### 🙂 Expected behavior

No lint issued, and `prop` typed as `'foo' | 'bar'`, as the type is well-formed; `keyof typeof DATA` uses only the keys of `DATA` and not its values, therefore `RefersToKey` is not a circular definition. Possible warning on the second definition, as the behaviour is unintuitive

### Additional information about the issue

My actual use-case involves data which makes reference to other parts of the data, via key names in the table. I want to be able to type-check that the values used are included in the data, using an automatically-generated key union type. A special type such as `auto` or `infer` would fulfil my use-case, if the following would hold:
```ts
const INFER_DEMO: Record<auto, number> = {
    foo: 1,
    bar: 2,
    baz: 3,
};
type Key = keyof typeof INFER_DEMO; // 'foo' | 'bar' | 'baz'
```

**Comment by andrewbranch** [comment]
> keyof typeof DATA is disjoint from RefersToKey; the keyof should erase the circular dependency

Unfortunately, we don’t have the ability to take the `keyof` a type that we’re still computing. `typeof DATA` has to be resolved first, which does depend on the `satisfies` expression, because the type there may contextually type the literals in the object.

> doesn't trip the lint, and - worse - typechecks successfully despite being wrong

This part is working as intended. `typeof DATA2` is `Record<any, DoesntTrip>` because that’s how you declared it. So the `keyof` that type is `any`. So the fully expanded type is `Record<any, { prop: any }>`.

**Comment by andrewbranch** [comment]
> A special type such as auto or infer would fulfil my use-case

This example really does look to be a good use for `satisfies`. In that simplified example, everything works:

```ts
const INFER_DEMO = {
    foo: 1,
    bar: 2,
    baz: 3,
} satisfies Record<any, number>;
type Key = keyof typeof INFER_DEMO; // 'foo' | 'bar' | 'baz'
```

The trouble only came when introducing a circularity, which would presumably be exactly the same with the imagined `auto` type.
